### PR TITLE
Smarter table/field name humanization

### LINKED ISSUE #2490: Be smarter about humanizing table and field names
Currently when "humanizing" table and field names we only split words on underscores and hyphens, e.x. `dashboard-card` and `dashboard_card` become `Dashboard Card`, but `dashboardcard`, `DashboardCard`, `dashboard card`, etc, do not. https://github.com/metabase/metabase/blob/master/src/metabase/models/common.clj#L109

At a minimum we should split on whitespace and probably lower-to-uppercase boundaries (e.x. `CamelCase` becomes `Camel Case`).

We could also attempt to detect individual words even within an all lowercase string with no delimiting characters using an algorithm such as http://stackoverflow.com/a/11642687 I tried that algorithm on a few of Metabase's own tables and it seems to work quite well:

```
dashboardcard
dashboard card

dashboardcardsubscription
dashboard card subscription

emailreport
email report

emailreportexecutions
email report executions
```

### PR DESCRIPTION
**Pull request by tlrobinson**
Resolves #2490 
1. split words on whitespace and lower-to-uppercase (i.e. camelCase) boundaries (in addition to underscore and dash)
2. infer spaces using a wordlist and fancy dynamic programming algorithm
3. remove adjacent duplicate words (e.x. "Query Query Execution" => "Query Execution")
4. uppercase "ID"

These rules are up for debate, but this works really well on Metabase's own application database (see examples below). We should also test it on a few other real databases.

The one incorrect case I've come across so far is `Liquibase` being converted to `Li Qui Base` ¯_(ツ)_/¯. That kind of thing can probably be solved by using a different or shorter word list (currently it's the top 125K works on English Wikipedia), or just whitelisting it.

| Old | New |
| --- | --- |
| Annotation Annotation | Annotation |
| Core Permissionsviolation | Core Permissions Violation |
| Core Userorgperm | Core User Org Perm |
| Dashboardcard Series | Dashboard Card Series |
| Databasechangelog | Database Change Log |
| Databasechangeloglock | Database Change Log Lock |
| Metabase Fieldvalues | Metabase Field Values |
| Metabase Foreignkey | Metabase Foreign Key |
| Metabase Tablesegment | Metabase Table Segment |
| Query Queryexecution | Query Execution |
| Report Cardfavorite | Report Card Favorite |
| Report Dashboardcard | Report Dashboard Card |
| Report Dashboardsubscription | Report Dashboard Subscription |
| Report Emailreport | Report Email Report |
| Report Emailreport Recipients | Report Email Report Executions |
| Report Emailreportexecutions | Report Email Report Recipients |
| Dateexecuted | Date Executed |
| Deployment Id | Deployment ID |
| Exectype | Exec Type |

### CODE REVIEW

**Comment by camsaul** [comment]
This is rad 😎

File: test/metabase/api/database_test.clj [hunk 1]
```diff
@@ -237,7 +237,7 @@
                                                  :table_id           (id :categories)
                                                  :special_type       "id"
                                                  :name               "ID"
-                                                 :display_name       "Id"
+                                                 :display_name       "ID"
```
Comment (agilliland) [inline test/metabase/api/database_test.clj]: is this desired?  we'd prefer to have `ID` than `Id`?

Comment (agilliland) [inline test/metabase/api/database_test.clj]: and just to be clear, i'm fine with it this way, i just want to make sure.  I assume we picked `Id` originally for a reason.

Comment (tlrobinson) [inline test/metabase/api/database_test.clj]: If by "we" you mean me, then yes :)

I think "ID" is correct outside of programming contexts. https://english.stackexchange.com/questions/101248/how-should-the-abbreviation-for-identifier-be-capitalized

Fun fact: apparently it originally stood for "identity document"

**Pull request merged**